src/groups/mqb/mqbs/mqbs_virtualstoragecatalog.cpp

```diff
@@ -448,42 +539,91 @@ int VirtualStorageCatalog::addVirtualStorage(bsl::ostream& errorDescription,
-        removeAll(appKey);
+    BSLS_ASSERT_SAFE(!appKey.isNull());
 
-        // Remove all virtual storages
-        d_virtualStorages.clear();
-        d_availableOrdinals.clear();
-        d_nextOrdinal = 0;
-        return true;  // RETURN
+    VirtualStoragesConstIter it = d_virtualStorages.findByKey2(appKey);
+    if (it == d_virtualStorages.end()) {
+        return mqbi::StorageResult::e_APPKEY_NOT_FOUND;
     }
 
-    VirtualStoragesConstIter it = d_virtualStorages.findByKey2(appKey);
-    if (it != d_virtualStorages.end()) {
-        // Make sure there is no AppMessage in the pending states
-        removeAll(appKey);
+    // Make sure there is no AppMessage in the pending states
 
-        const VirtualStorage& vs = *it->value();
-        d_availableOrdinals.insert(vs.ordinal());
+    BSLS_ASSERT_SAFE(!d_ordinals.empty());
 
-        if (d_queue_p) {
-            BSLS_ASSERT_SAFE(d_queue_p->queueEngine());
-            // QueueEngines use the key to look up the id
+    VirtualStorage*    removing        = it->value().get();
+    const unsigned int removingOrdinal = removing->ordinal();
 
-            d_queue_p->queueEngine()->unregisterStorage(vs.appId(),
-                                                        appKey,
-                                                        vs.ordinal());
+    VirtualStorageSp replacing;
+    unsigned int     replacingOrdinal = removingOrdinal;
+
+    if (d_areOrdinalsContinuous) {
+        replacing        = d_ordinals.back();
+        replacingOrdinal = replacing->ordinal();
+        BSLS_ASSERT_SAFE(replacingOrdinal + 1 == d_ordinals.size());
+    }
+
+    BALL_LOG_INFO << "Removing VirtualStorage with appId '"
+                  << removing->appId() << "' & appKey '" << removing->appKey()
+                  << "' for queue '" << d_storage_p->queueUri()
+                  << "' & queueKey '" << d_storage_p->queueKey()
+                  << "'. Replacing " << removingOrdinal << " with "
+                  << replacingOrdinal;
+
+    // Replace [vs->ordinal()] with [maxVirtualStorageSpOrdinal]
+
+    DataStreamIterator itData;
+    bsls::Types::Int64 total = d_dataStream.size();
+
+    if (seek(&itData, removing) < total) {
+        if (cb) {
+            mqbi::StorageResult::Enum rc = cb(appKey, itData);
+            if (mqbi::StorageResult::e_SUCCESS != rc) {
+                return rc;
+            }
         }
+    }
 
-        d_virtualStorages.erase(it);
-        return true;  // RETURN
+    purgeImpl(removing, itData, replacingOrdinal);
+
+    if (d_queue_p) {
+        BSLS_ASSERT_SAFE(d_queue_p->queueEngine());
+        // QueueEngines use the key to look up the id
+
+        d_queue_p->queueEngine()->unregisterStorage(removing->appId(),
+                                                    appKey,
+                                                    removingOrdinal);
     }
 
-    return false;
+    if (d_areOrdinalsContinuous) {
```

Comment:
I'm fairly sure here that the code is more readable with `d_isProxy` or such.  In my mind, at least, when reading this component, that's what I've found myself reading it as, and the documentation comment says the same.  Willing to be overridden, though.